### ISSUE TITLE
An algorithm used to break/detect cycles between constant fields exhibits non-deterministic behavior and might produce unexpected diagnostics

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#50127]
The following unit-test succeeds today:
```
        [Fact]
        public void Test_01()
        {
            var source =
@"enum C
{
    A = F,
    B = A + C + D + E,
    C = E,
    D = B,
    E = B,
    F = G,
    G = F,
}";

            System.Collections.Immutable.ImmutableArray<Diagnostic> listOf2 = default;
            System.Collections.Immutable.ImmutableArray<Diagnostic> listOf3 = default;

            for (int i = 0; i < 100; i++)
            {
                var compilation = CreateCompilation(source);
                var list = compilation.GetDeclarationDiagnostics();

                switch (list.Length)
                {
                    case 2:
                        listOf2 = list;
                        break;
                    case 3:
                        listOf3 = list;
                        break;
                    default:
                        Assert.True(false);
                        break;
                }

                if (!listOf2.IsDefault && !listOf3.IsDefault)
                {
                    break;
                }
            }

            listOf2.Verify(
                // (4,5): error CS0110: The evaluation of the constant value for 'C.B' involves a circular definition
                //     B = A + C + D + E,
                Diagnostic(ErrorCode.ERR_CircConstValue, "B").WithArguments("C.B").WithLocation(4, 5),
                // (8,5): error CS0110: The evaluation of the constant value for 'C.F' involves a circular definition
                //     F = G,
                Diagnostic(ErrorCode.ERR_CircConstValue, "F").WithArguments("C.F").WithLocation(8, 5)
                );

            listOf3.Verify(
                // (3,5): error CS0110: The evaluation of the constant value for 'C.A' involves a circular definition
                //     A = F,
                Diagnostic(ErrorCode.ERR_CircConstValue, "A").WithArguments("C.A").WithLocation(3, 5),
                // (4,5): error CS0110: The evaluation of the constant value for 'C.B' involves a circular definition
                //     B = A + C + D + E,
                Diagnostic(ErrorCode.ERR_CircConstValue, "B").WithArguments("C.B").WithLocation(4, 5),
                // (8,5): error CS0110: The evaluation of the constant value for 'C.F' involves a circular definition
                //     F = G,
                Diagnostic(ErrorCode.ERR_CircConstValue, "F").WithArguments("C.F").WithLocation(8, 5)
                );
        }
```

Observed:
The same code sometimes produces 2 errors and sometimes 3 errors. The error that appears/disappears is
```
                // (3,5): error CS0110: The evaluation of the constant value for 'C.A' involves a circular definition
                //     A = F,
                Diagnostic(ErrorCode.ERR_CircConstValue, "A").WithArguments("C.A").WithLocation(3, 5),
```
It is also somewhat misleading because ```C.A``` is not involved into a cycle itself, it depends on a field that is involved into a cycle that is reported separately.

I believe the intent was to report the error only for fields that are part of a cycle.

**Comment by Youssef1313** [comment]
Similar issue: https://github.com/dotnet/roslyn-analyzers/issues/4544 (might not be related)
